src/main/java/gregtech/api/util/InventoryUtils.java

```diff
@@ -90,29 +131,93 @@ public static boolean simulateItemStackMerge(List<ItemStack> items,
      * @param source      the ItemStacks to merge into {@code destination}.
      * @param destination a target inventory of existing ItemStacks.
      */
-    private static void mergeItemStacks(Collection<ItemStack> source, Collection<ItemStack> destination) {
+    static void mergeItemStacks(Collection<ItemStack> source, Collection<ItemStack> destination)
+    {
         // Since we're mutating the collection during iteration, use an iterator.
         final Iterator<ItemStack> sourceItemStacks = source.iterator();
-        while(sourceItemStacks.hasNext()) {
+        while(sourceItemStacks.hasNext())
+        {
             final ItemStack sourceItemStack = sourceItemStacks.next();
 
             // Find a matching item in the output bus, if any
             for(ItemStack destItemStack : destination)
-                if(ItemStack.areItemsEqual(destItemStack, sourceItemStack)) {
+                if(ItemStack.areItemsEqual(destItemStack, sourceItemStack) &&
+                   ItemStack.areItemStackTagsEqual(destItemStack, sourceItemStack))
+                {
                     // if it's possible to merge stacks
                     final int availableSlots = destItemStack.getMaxStackSize() - destItemStack.getCount();
-                    if(availableSlots > 0) {
+                    if(availableSlots > 0)
+                    {
                         final int itemCount = Math.min(availableSlots, sourceItemStack.getCount());
                         sourceItemStack.shrink(itemCount);
                         destItemStack.grow(itemCount);
 
                         // if the output stack was merged completely, remove it and stop looking
-                        if(sourceItemStack.isEmpty()) {
+                        if(sourceItemStack.isEmpty())
+                        {
                             sourceItemStacks.remove();
                             break;
                         }
                     }
                 }
         }
     }
+
+    /**
+     * Normalizes a potentially problematic ItemStack by splitting it into copied stacks of at most
+     * {@link ItemStack#getMaxStackSize()} items. Empty ItemStacks produce an empty
+     * result list, and stacks already in normalized form are copied as a singleton list.
+     *
+     * @param stack an ItemStack to process.
+     * @return an immutable List of the resulting ItemStacks, in descending size order.
+     */
+    public static List<ItemStack> normalizeItemStack(ItemStack stack)
```

Comment:
I think it would be worth it even apply this function to all places which could use it as port of this PR.

For example https://github.com/GregTechCE/GregTech/blob/master/src/main/java/gregtech/common/metatileentities/multi/electric/MetaTileEntityMultiFurnace.java#L217

Reply:
I've been working on adding this change since you requested it, haven't committed that just yet though

Reply:
Let's get this done in another PR so I can get this in current release

Reply:
All right, that's fine. I can move it to a new feature branch and PR it in later, then.